File: typescript/packages/bench/src/contracts/metricContract.ts
/** Single-number contract. */

import type { Dataset } from "../dataset";

/** What a metric's number means. */
export enum MetricUnit {
    Count = "count",
    Cents = "cents",
    Percent = "percent",
    Days = "days",
}

// eslint-disable-next-line @typescript-eslint/no-namespace
export namespace MetricUnit {
    const SUFFIXES: Readonly<Record<MetricUnit, string>> = {
        [MetricUnit.Count]: "",
        [MetricUnit.Cents]: " c",
        [MetricUnit.Percent]: "%",
        [MetricUnit.Days]: " d",
    };

    export function suffix(unit: MetricUnit): string {
        return SUFFIXES[unit];
    }
}

/** One scalar computed over the whole dataset. */
export interface MetricContract {
    /** Registry key, unique across the lane. */
    readonly key: string;

    readonly unit: MetricUnit;

    /** The number itself. */
    compute(data: Dataset): number;

    /** Display form with the unit suffix appended. */
    formatted(data: Dataset): string;
}

File: rust/crates/atelier-bench/src/contracts/metric_contract.rs
//! Single-number contract.

use crate::dataset::Dataset;

/// What a metric's number means.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum MetricUnit {
    Count,
    Cents,
    Percent,
    Days,
}

impl MetricUnit {
    pub fn suffix(&self) -> &'static str {
        match self {
            MetricUnit::Count => "",
            MetricUnit::Cents => " c",
            MetricUnit::Percent => "%",
            MetricUnit::Days => " d",
        }
    }
}

/// One scalar computed over the whole dataset.
pub trait MetricContract: Send + Sync {
    /// Registry key, unique across the lane.
    fn key(&self) -> &'static str;

    fn unit(&self) -> MetricUnit;

    /// The number itself.
    fn compute(&self, data: &Dataset) -> f64;

    /// Display form with the unit suffix appended.
    fn formatted(&self, data: &Dataset) -> String {
        format!("{:.2}{}", self.compute(data), self.unit().suffix())
    }
}
